actionpack/lib/action_controller/metal/redirecting.rb

```diff
@@ -92,10 +98,28 @@ def redirect_back(fallback_location:, allow_other_host: true, **args)
     #
     # All other options that can be passed to #redirect_to are accepted as
     # options and the behavior is identical.
-    def redirect_back_or_to(fallback_location, allow_other_host: true, **args)
-      referer = request.headers["Referer"]
-      redirect_to_referer = referer && (allow_other_host || _url_host_allowed?(referer))
-      redirect_to redirect_to_referer ? referer : fallback_location, **args
+    def redirect_back_or_to(fallback_location, allow_other_host: allow_open_redirects, **args)
+      referer = request.headers.fetch("Referer", fallback_location)
+      response_options = {
+        fallback_location: allow_open_redirects ? fallback_location : nil,
+        allow_other_host: allow_other_host,
+        **args,
+      }
+      redirect_to referer, response_options
+    end
+
+    def _compute_safe_redirect_to_location(request, options, response_options)
+      location = _compute_redirect_to_location(request, options)
+      if response_options[:allow_other_host] || _url_host_allowed?(location)
+        location
+      else
+        fallback_location = response_options[:fallback_location] || raise(ArgumentError, <<~MSG.squish)
+          Unsafe redirect #{location.inspect},
+          use :fallback_location to specify a fallback
+          or :allow_other_host to redirect anyway.
+        MSG
+        _compute_redirect_to_location(request, fallback_location)
+      end
```

Comment:
This method feels too complex. Let's see if the other changes allow for slimming here otherwise I'll check back and suggest something on the next pass.

Reply:
I moved the `raise` to a separate method, but feel free to make further suggestions 😄